### PR TITLE
[Merged by Bors] - feat(Algebra/Star/StarRingHom): Add non-unital star ring homomorphisms

### PR DESCRIPTION
**Pull request by mans0954**
Definitions and basic properties of  non-unital star ring homomorphisms

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by j-loreaux** [comment]
no need to include me in the copyright.

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+section
+
+variable {A B C : Type*}
+variable [NonUnitalNonAssocSemiring A] [Star A]
+variable [NonUnitalNonAssocSemiring B] [Star B]
+variable [NonUnitalNonAssocSemiring C] [Star C]
+
+instance : FunLike (A →⋆ₙ+* B) A B
+    where
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:67]:
```suggestion
instance : FunLike (A →⋆ₙ+* B) A B where
```

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  coe f := f.toFun
+  coe_injective' := by rintro ⟨⟨⟨f, _⟩,  _⟩, _⟩ ⟨⟨⟨g, _⟩, _⟩, _⟩ h; congr
+
+instance : NonUnitalRingHomClass (A →⋆ₙ+* B) A B
+    where
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:72]:
```suggestion
instance : NonUnitalRingHomClass (A →⋆ₙ+* B) A B where
```

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+  map_mul f := f.map_mul'
+  map_add f := f.map_add'
+  map_zero f := f.map_zero'
+
+instance : NonUnitalStarRingHomClass (A →⋆ₙ+* B) A B
+    where
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:78]:
```suggestion
instance : NonUnitalStarRingHomClass (A →⋆ₙ+* B) A B where
```

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  map_star : ∀ (f : F) (a : A), f (star a) = star (f a)
+
+namespace StarRingEquiv
+
+section Basic
+
+variable {A B : Type*} [Add A] [Add B] [Mul A] [Mul B] [Star A] [Star B]
+
+instance : EquivLike (A ≃⋆+* B) A B
+    where
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:134]:
```suggestion
instance : EquivLike (A ≃⋆+* B) A B where
```

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+  inv f := f.invFun
+  left_inv f := f.left_inv
+  right_inv f := f.right_inv
+  coe_injective' f g h₁ h₂ := by
+    rcases f with ⟨⟨⟨_, _, _⟩, _⟩, _⟩
+    rcases g with ⟨⟨⟨_, _, _⟩, _⟩, _⟩
+    congr
+
+instance : RingEquivClass (A ≃⋆+* B) A B
+    where
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:145]:
```suggestion
instance : RingEquivClass (A ≃⋆+* B) A B where
```

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+  map_mul f := f.map_mul'
+  map_add f := f.map_add'
+
+instance : StarRingEquivClass (A ≃⋆+* B) A B
+    where
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:150]:
```suggestion
instance : StarRingEquivClass (A ≃⋆+* B) A B where
```

**Review by j-loreaux — COMMENTED** [review]
Looks like a decent start, but still a lot of the API is missing. It's okay if you add it in follow-up PRs though.

Also, when you're done with this, would you mind fixing up [StarRingHomClass.instOrderHomClass](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Star/Order.html#StarRingHomClass.instOrderHomClass) and [StarRingHomClass.instOrderIsoClass](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Star/Order.html#StarRingHomClass.instOrderIsoClass) to use these new classes (and renaming the latter?

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+/-- A *⋆-ring* equivalence is an equivalence preserving addition, multiplication, and the star
+operation, which allows for considering both unital and non-unital equivalences with a single
+structure. -/
+structure StarRingEquiv (A B : Type*) [Add A] [Add B] [Mul A] [Mul B] [Star A] [Star B]
+    extends A ≃+* B where
+  /-- By definition, a ⋆-algebra equivalence preserves the `star` operation. -/
+  map_star' : ∀ a : A, toFun (star a) = star (toFun a)
+
+@[inherit_doc StarRingEquiv] infixr:25 " ≃⋆+* " => StarRingEquiv _
+
```
[the remark this answers is not in the document]
Comment (mans0954) [reply Mathlib/Algebra/Star/StarRingHom.lean:106]: `StarAlgEquiv` has this https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/Algebra/Star/StarAlgHom.lean#L713 - why do we not need the equivalent here?

[the remark this answers is not in the document]
Comment (j-loreaux) [reply Mathlib/Algebra/Star/StarRingHom.lean:164]: Because there's no scalar ring to possibly leave to unification.

**Comment by mans0954** [comment]
> Looks like a decent start, but still a lot of the API is missing. It's okay if you add it in follow-up PRs though.

I've fleshed out the API now.

> 
> Also, when you're done with this, would you mind fixing up [StarRingHomClass.instOrderHomClass](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Star/Order.html#StarRingHomClass.instOrderHomClass) and [StarRingHomClass.instOrderIsoClass](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Star/Order.html#StarRingHomClass.instOrderIsoClass) to use these new classes (and renaming the latter?

I've done this in a separate PR #13089.

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+      { (⟨⟨⟨f, f', h₁, h₂⟩, h₃, h₄⟩, h₅⟩ : A ≃⋆+* B).symm with
+        toFun := f'
+        invFun := f } :=
+  rfl
+
+@[simp]
+theorem refl_symm : (StarRingEquiv.refl : A ≃⋆+* A).symm = StarRingEquiv.refl :=
+  rfl
+
+/-- Star ring equivalences are transitive. -/
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:389]: Ditto

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+  ext
+  rfl
+
+theorem symm_bijective : Function.Bijective (symm : (A ≃⋆+* B) → B ≃⋆+* A) :=
+  Function.bijective_iff_has_inverse.mpr ⟨_, symm_symm, symm_symm⟩
+
+@[simp]
+theorem mk_coe' (e : A ≃⋆+* B) (f h₁ h₂ h₃ h₄ h₅) :
+    (⟨⟨⟨f, e, h₁, h₂⟩, h₃, h₄⟩, h₅⟩ : B ≃⋆+* A) = e.symm :=
+  symm_bijective.injective <| ext fun _ => rfl
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:375]: I really don't think this is the theorem we want. It's unwrapping too many structures, and why `symm`?

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+
+@[simp]
+theorem symm_mk (f f') (h₁ h₂ h₃ h₄ h₅) :
+    (⟨⟨⟨f, f', h₁, h₂⟩, h₃, h₄⟩, h₅⟩ : A ≃⋆+* B).symm =
+      { (⟨⟨⟨f, f', h₁, h₂⟩, h₃, h₄⟩, h₅⟩ : A ≃⋆+* B).symm with
+        toFun := f'
+        invFun := f } :=
+  rfl
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:383]: Same as above, but also I think the naming is backwards between these two.

**Review by j-loreaux — COMMENTED** [review]
I'll probably have some more to say after this, but I wanted to get the easy stuff out of the way.

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+  { hF with }
+
+-- See note [lower instance priority]
+instance (priority := 100) {F A B : Type*} {_ : NonUnitalNonAssocSemiring A} {_ : Star A}
+    {_ : NonUnitalNonAssocSemiring B} {_ : Star B} [EquivLike F A B] [RingEquivClass F A B]
+    [StarRingEquivClass F A B] : NonUnitalStarRingHomClass F A B :=
+  { }
+
+/-- Turn an element of a type `F` satisfying `StarAlgEquivClass F R A B` into an actual
+`StarAlgEquiv`. This is declared as the default coercion from `F` to `A ≃⋆ₐ[R] B`. -/
```
[the remark this answers is not in the document]
Comment (mans0954) [reply Mathlib/Algebra/Star/StarRingHom.lean]: Done, thanks.

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 12]
```diff
@@ -0,0 +1,467 @@
...
+  rfl
+
+@[ext]
+theorem ext {f g : A ≃⋆+* B} (h : ∀ a, f a = g a) : f = g :=
+  DFunLike.ext f g h
+
+theorem ext_iff {f g : A ≃⋆+* B} : f = g ↔ ∀ a, f a = g a :=
+  DFunLike.ext_iff
+
+/-- Star ring equivalences are reflexive. -/
```
[the remark this answers is not in the document]
Comment (mans0954) [reply Mathlib/Algebra/Star/StarRingHom.lean:327]: This is _Mutatis mutandis_ from https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/Algebra/Star/StarAlgHom.lean#L846

File: Mathlib/Algebra/Star/StarRingHom.lean (hunk 9 above)
Comment (mans0954) [reply Mathlib/Algebra/Star/StarRingHom.lean:372]: This is _mutatis mutandis_ from  https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/Algebra/Star/StarAlgHom.lean#L904

File: Mathlib/Algebra/Star/StarRingHom.lean (hunk 12 above)
[the remark this answers is not in the document]
Comment (j-loreaux) [reply Mathlib/Algebra/Star/StarRingHom.lean:327]: Wow, I can't believe I wrote that.

File: Mathlib/Algebra/Star/StarRingHom.lean (hunk 9 above)
Comment (j-loreaux) [reply Mathlib/Algebra/Star/StarRingHom.lean:372]: Both need to be fixed then. It's likely a consequence of the port.

Comment (mans0954) [reply Mathlib/Algebra/Star/StarRingHom.lean:372]: Mathlib3 algbera version: https://github.com/leanprover-community/mathlib/blob/master/src/algebra/star/star_alg_hom.lean#L619
```
@[simp] lemma mk_coe' (e : A ≃⋆ₐ[R] B) (f h₁ h₂ h₃ h₄ h₅ h₆) :
  (⟨f, e, h₁, h₂, h₃, h₄, h₅, h₆⟩ : B ≃⋆ₐ[R] A) = e.symm :=
symm_bijective.injective $ ext $ λ x, rfl
```

File: Mathlib/Algebra/Star/StarRingHom.lean (hunk 12 above)
[the remark this answers is not in the document]
Comment (mans0954) [reply Mathlib/Algebra/Star/StarRingHom.lean]: #13266

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 13]
```diff
@@ -0,0 +1,467 @@
...
+#align star_alg_equiv.refl StarRingEquiv.refl
+
+instance : Inhabited (A ≃⋆+* A) :=
+  ⟨refl⟩
+
+@[simp]
+theorem coe_refl : ⇑(refl : A ≃⋆+* A) = id :=
+  rfl
+
+/-- Star ring equivalences are symmetric. -/
```
[the remark this answers is not in the document]
Comment (mans0954) [reply Mathlib/Algebra/Star/StarRingHom.lean]: #13266

File: Mathlib/Algebra/Star/StarRingHom.lean (hunk 8 above)
Comment (mans0954) [reply Mathlib/Algebra/Star/StarRingHom.lean]: #13266

File: Mathlib/Algebra/Star/StarRingHom.lean (hunk 9 above)
Comment (mans0954) [reply Mathlib/Algebra/Star/StarRingHom.lean]: Perhaps we could just deprecate them? #13267

File: Mathlib/Algebra/Star/StarRingHom.lean (hunk 10 above)
Comment (mans0954) [reply Mathlib/Algebra/Star/StarRingHom.lean]: Perhaps we could just deprecate them? #13267

File: Mathlib/Algebra/Star/StarRingHom.lean (hunk 9 above)
Comment (mans0954) [reply Mathlib/Algebra/Star/StarRingHom.lean]: @j-loreaux now #13267 has been queued for merging, I've made the corresponding update to this PR.

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 14]
```diff
@@ -0,0 +1,469 @@
...
+Authors: Christopher Hoskin
+-/
+import Mathlib.Algebra.Ring.Defs
+import Mathlib.Algebra.Star.Basic
+
+/-!
+# Morphisms of star rings
+
+This file defines morphisms between (non-unital) rings `A` and `B` where both
+`A` and `B` are equipped with a `star` operation. This morphisms, namely `NonUnitalStarRingHom` is
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:13]:
```suggestion
This file defines a new type of morphism between (non-unital) rings `A` and `B` where both
`A` and `B` are equipped with a `star` operation. This morphism, namely `NonUnitalStarRingHom`, is
```

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 15]
```diff
@@ -0,0 +1,469 @@
...
+open EquivLike
+
+/-! ### Non-unital star ring homomorphisms -/
+
+/-- A *non-unital ⋆-ring homomorphism* is a non-unital ring homomorphism between non-unital
+non-associative semirings `A` and `B` equipped with a `star` operation, and this homomorphism is
+also `star`-preserving. -/
+structure NonUnitalStarRingHom (A B : Type*) [NonUnitalNonAssocSemiring A]
+  [Star A] [NonUnitalNonAssocSemiring B]
+  [Star B] extends A →ₙ+* B where
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:41]:
```suggestion
    [Star A] [NonUnitalNonAssocSemiring B] [Star B] extends A →ₙ+* B where
```

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 16]
```diff
@@ -0,0 +1,469 @@
...
+  /-- By definition, a non-unital ⋆-ring homomorphism preserves the `star` operation. -/
+  map_star' : ∀ a : A, toFun (star a) = star (toFun a)
+
+/-- `α →⋆ₙ+* β` denotes the type of non-unital ring homomorphisms from `α` to `β`. -/
+infixr:25 " →⋆ₙ+* " => NonUnitalStarRingHom
+
+/-- Reinterpret a non-unital star ring homomorphism as a non-unital ring homomorphism
+by forgetting the interaction with the star operation. -/
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:49]:
```suggestion
/-- Reinterpret a non-unital star ring homomorphism as a non-unital ring homomorphism
by forgetting the interaction with the star operation. 

Users should not make use of this, but instead utilize the coercion obtained through
the `NonUnitalRingHomClass` instance. -/
```

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 17]
```diff
@@ -0,0 +1,469 @@
...
+
+/-- See Note [custom simps projection] -/
+def Simps.apply (f : A →⋆ₙ+* B) : A → B := f
+
+initialize_simps_projections NonUnitalStarRingHom (toFun → apply)
+
+@[simp]
+protected theorem coe_coe {F : Type*} [FunLike F A B] [NonUnitalRingHomClass F A B]
+    [NonUnitalStarRingHomClass F A B] (f : F) :
+    ⇑(f : A →⋆ₙ+* B) = f := rfl
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:107]:
```suggestion
    [NonUnitalStarRingHomClass F A B] (f : F) : ⇑(f : A →⋆ₙ+* B) = f :=
  rfl
```

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 18]
```diff
@@ -0,0 +1,469 @@
...
+infixr:25 " →⋆ₙ+* " => NonUnitalStarRingHom
+
+/-- Reinterpret a non-unital star ring homomorphism as a non-unital ring homomorphism
+by forgetting the interaction with the star operation. -/
+add_decl_doc NonUnitalStarRingHom.toNonUnitalRingHom
+
+/-- `NonUnitalStarRingHomClass F A B` states that `F` is a type of non-unital ⋆-ring homomorphisms.
+You should also extend this typeclass when you extend `NonUnitalStarRingHom`. -/
+class NonUnitalStarRingHomClass (F : Type*) (A B : outParam Type*)
+     [NonUnitalNonAssocSemiring A] [Star A] [NonUnitalNonAssocSemiring B] [Star B]
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:55]:
```suggestion
    [NonUnitalNonAssocSemiring A] [Star A] [NonUnitalNonAssocSemiring B] [Star B]
```

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 19]
```diff
@@ -0,0 +1,469 @@
...
+operation. -/
+add_decl_doc StarRingEquiv.toRingEquiv
+
+/-- `StarRingEquivClass F A B` asserts `F` is a type of bundled ⋆-ring equivalences between `A` and
+`B`.
+You should also extend this typeclass when you extend `StarRingEquiv`. -/
+class StarRingEquivClass (F : Type*) (A B : outParam Type*)
+  [Add A] [Mul A] [Star A] [Add B] [Mul B] [Star B] [EquivLike F A B]
+  extends RingEquivClass F A B : Prop
+  where
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:254]:
```suggestion
    [Add A] [Mul A] [Star A] [Add B] [Mul B] [Star B] [EquivLike F A B]
    extends RingEquivClass F A B : Prop where
```

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 20]
```diff
@@ -0,0 +1,469 @@
...
+  /-- By definition, a ⋆-ring equivalence preserves the `star` operation. -/
+  map_star : ∀ (f : F) (a : A), f (star a) = star (f a)
+
+namespace StarRingEquivClass
+
+-- See note [lower instance priority]
+instance (priority := 50) {F A B : Type*} {_ : Add A} {_ : Mul A} {_ : Star A} {_ : Add B}
+    {_ : Mul B} {_ : Star B} [EquivLike F A B] [hF : StarRingEquivClass F A B] :
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:262]: @eric-wieser do we still need these parameters to be implicit, or is this now a porting artifact? I'm inclined to think it's the latter.

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 21]
```diff
@@ -0,0 +1,469 @@
...
+
+@[simp]
+theorem refl_symm : (StarRingEquiv.refl : A ≃⋆+* A).symm = StarRingEquiv.refl :=
+  rfl
+
+/-- Transitivity of `StarRingEquiv`. -/
+@[trans]
+def trans (e₁ : A≃⋆+* B) (e₂ : B ≃⋆+* C) : A ≃⋆+* C :=
+  { e₁.toRingEquiv.trans
+      e₂.toRingEquiv with
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:395]:
```suggestion
  { e₁.toRingEquiv.trans e₂.toRingEquiv with
```

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 22]
```diff
@@ -0,0 +1,469 @@
...
+  right_inv := h₂
+  map_add' := map_add f
+  map_mul' := map_mul f
+  map_star' := map_star f
+
+/-- Promote a bijective star ring homomorphism to a star ring equivalence. -/
+noncomputable def ofBijective (f : F) (hf : Function.Bijective f) : A ≃⋆+* B :=
+  {
+    RingEquiv.ofBijective f
+      (hf : Function.Bijective (f : A → B)) with
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:454]:
```suggestion
  { RingEquiv.ofBijective f (hf : Function.Bijective (f : A → B)) with
```

File: Mathlib/Algebra/Star/StarRingHom.lean [hunk 23]
```diff
@@ -0,0 +1,469 @@
...
+variable [FunLike F A B] [NonUnitalRingHomClass F A B]
+
+/-- Turn an element of a type `F` satisfying `NonUnitalStarRingHomClass F A B` into an actual
+`NonUnitalStarRingHom`. This is declared as the default coercion from `F` to `A →⋆ₙ+ B`. -/
+@[coe]
+def toNonUnitalStarRingHom [NonUnitalStarRingHomClass F A B] (f : F) : A →⋆ₙ+* B :=
+  { (f : A →ₙ+* B) with
+    map_star' := map_star f }
+
+instance [NonUnitalStarRingHomClass F A B] : CoeTC F (A →⋆ₙ+* B) :=
```
Comment (j-loreaux) [inline Mathlib/Algebra/Star/StarRingHom.lean:72]: I *think* this should be a `CoeHead` instance. If Lean doesn't complain, let's go with that.
```suggestion
instance [NonUnitalStarRingHomClass F A B] : CoeHead F (A →⋆ₙ+* B) :=
```

**Review by j-loreaux — COMMENTED** [review]
Looks pretty good. I'm ready to merge subject to these comments and Eric's response to my question above.

File: Mathlib/Algebra/Star/StarRingHom.lean (hunk 20 above)
Comment (mans0954) [reply Mathlib/Algebra/Star/StarRingHom.lean:263]: Link to the porting note in the porting PR by @ADedecker : https://github.com/leanprover-community/mathlib4/pull/2603/files#diff-8017bfdc9f21755fb88462d9ba5909d1590cfe91939577ce589ef704a831df57R700

Comment (mans0954) [reply Mathlib/Algebra/Star/StarRingHom.lean]: Eric is okay with https://github.com/leanprover-community/mathlib4/pull/14110#pullrequestreview-2140003696 so I've made the corresponding changes here.

**Comment by j-loreaux** [comment]
bors merge

**Pull request closed without merging**